### PR TITLE
Custom step range picker for funnels time conversion

### LINKED ISSUE #5701: Custom step range selection in time conversion graph
## Is your feature request related to a problem?

We should allow for custom step range selection in our time conversion graphs. Currently we only allow for adjacent step selection.

Screen Shot 2021-08-23 at 5 13 14 PM

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

Closes #5701.

Screen Shot 2021-08-24 at 10 41 15 AM

- [x] Give user ability to choose step ranges > 1 step long
- [x] Refactor step selector out into separate component
- [x] Consolidate data source for `histogramSteps` into `filters` in logic

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious
- [ ] New/changed UI is decent on smartphones (viewport width around 360px)

### CODE REVIEW

**Review by Twixes — DISMISSED** [review]
Thank you for tackling this!
I have 2 major comments:
Screen Shot 2021-08-24 at 12 28 44
1. This is pretty awkwardly placed and reduces area available to the chart below. I think the Options pane may be the most natural place for these controls. Also happy to hear from @clarkus regarding UX.
2. The feature should be available for both Time to convert and Historical trends (the API is identical).

**Comment by alexkim205** [comment]
Thanks for taking a look :) agreed options might be the best place for this

**Comment by alexkim205** [comment]
How's this look? 

Screen Shot 2021-08-24 at 10 41 15 AM

**Comment by clarkus** [comment]
Agree it's best placed under options for now.

**Comment by liyiy** [comment]
When I change it from days to hours or weeks or minutes nothing happens :o 

https://user-images.githubusercontent.com/25164963/130728097-ddbda800-b65e-45cf-b6f8-561d92dc135e.mov

**Comment by mariusandra** [comment]
When looking at the interface, I have a hard time understanding what "between step 1 and step 3" does, and what kind of side effects it has.

I understand what is "14 days conversion window", but if I have 4 steps and set "14 days between steps 1 and 2", what happens to the other steps? They can take forever? 

I'd also reorder the two options. The one about conversion window should probably be first, as I imagine that's the one that'll be interacted with more.

**Comment by neilkakkar** [comment]
Wait, wait, what? (Marius's comment made me look more carefully into this, so thanks for this). This is pretty confusing, agree!

We aren't changing the conversion window sizes between steps here - this isn't possible.

What we're changing is the `Steps` option on the Time to Convert graph ~~only~~. So, this shouldn't show up for ~~historical trends, nor~~ for the ordinary funnel.

 I think this should stay as far away from the conversion window option as possible, since it's unrelated. Thoughts?

**Comment by neilkakkar** [comment]
To add a bit to the confusion, just looked at how they're used in HistoricalTrends vs TimeToConvert, and the semantics differ slightly:

1. In Trends, looking between Step X and Step Y means you're looking at the **conversion rate** from Step X to Step Y.
2. In TimeToConvert, looking between Step X and Step Y means you're looking at the **time taken to convert** between those two steps.

While this can be obvious to someone entrenched in the 2 features and their usage, seems non-trivial to infer for newbies.

This might impact how we display the copy^

File: frontend/src/scenes/funnels/funnelLogic.ts [hunk 1]
```diff
@@ -170,7 +169,10 @@ export const funnelLogic = kea<funnelLogicType>({
             breakdown_value,
         }),
         setStepReference: (stepReference: FunnelStepReference) => ({ stepReference }),
-        changeHistogramStep: (from_step: number, to_step: number) => ({ from_step, to_step }),
+        changeHistogramStep: (funnel_from_step?: number, funnel_to_step?: number) => ({
```
Comment (Twixes) [inline frontend/src/scenes/funnels/funnelLogic.ts:172]: Would be good to also remove the `histogram` characterization in this logic to something more general, maybe `changeStepRange` here for instance.

File: frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelStepsPicker.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        return null
+    }
+
+    const onChange = (funnel_from_step?: number, funnel_to_step?: number): void => {
+        changeHistogramStep(funnel_from_step, funnel_to_step)
+    }
+
+    return (
+        <Row align="middle" style={{ margin: '8px 0' }}>
+            Between
```
Comment (Twixes) [inline frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelStepsPicker.tsx:22]: I think @mariusandra and @neilkakkar have a point. How about renaming this from just "Between X and Y" to:
- for Time to convert: "Conversion time from Step X to Step Y",
- for Historical trends: "Conversion rate from Step X to Step Y"?

**Review by Twixes — DISMISSED** [review]

**Comment by alexkim205** [comment]
> When I change it from days to hours or weeks or minutes nothing happens :o

Hmm I believe you'll have to unfocus from the input to trigger a change bc we saw that triggering onChange would be expensive. Does that work?

> I think this should stay as far away from the conversion window option as possible, since it's unrelated. Thoughts?

I can see how people would think that conversion windows and step ranges are interrelated if they're located in the same space, and we definitely don't want to cause any confusion. Maybe it'd make more sense to move this into the steps tab?

**Comment by liyiy** [comment]
> > When I change it from days to hours or weeks or minutes nothing happens :o
> 
> Hmm I believe you'll have to unfocus from the input to trigger a change bc we saw that triggering onChange would be expensive. Does that work?

I feel like I wouldn't know to do that as a user though, can we make it unfocus automatically? And add a breakpoint if the user is changing the options too quickly

**Comment by alexkim205** [comment]
> I feel like I wouldn't know to do that as a user though, can we make it unfocus automatically? And add a breakpoint if the user is changing the options too quickly

Good idea! 👍

**Comment by alexkim205** [comment]
Pinging @Twixes and @neilkakkar for rereview. Updated copy and added a tooltip. Open to suggestions for rewording

Time to convert

Screen Shot 2021-08-25 at 10 10 05 AM

Trends

Screen Shot 2021-08-25 at 10 11 48 AM

**Comment by clarkus** [comment]
I caught up with @alexkim205 via slack to talk through this issue. I gave poor guidance earlier as I misunderstood the problem being solved. 

I think we should change placement. Each scenario is specific to the graph type selected, and it directly modifies the metric in the header (conversion time or conversion rate) based on the steps configured. We can place a configuration control there or just build the step selectors into the header for the visualization. In the example below you can see repeated headers for each scenario: conversion steps, time to convert, and historical trends. Thoughts?

Screen Shot 2021-08-25 at 11 39 24 AM

https://www.figma.com/file/9yWtngNb1AIuf6KmXaEPJA/App-doodles?node-id=1139%3A135838

**Comment by neilkakkar** [comment]
@alexkim205 the recent change is better, but I really do love Chris's suggestion! (so far away from conversion window, yay 😂) - makes the most sense to me.

**Comment by alexkim205** [comment]
Done!

Steps

Screen Shot 2021-08-25 at 2 05 41 PM


Convert

Screen Shot 2021-08-25 at 2 03 02 PM

Trends

Screen Shot 2021-08-25 at 2 02 57 PM

tagging @mariusandra and @neilkakkar and @clarkus for rereview

**Pull request merged**